manifests/init.pp

```diff
@@ -51,73 +51,53 @@
 # Copyright 2013 Thomas Linkin, unless otherwise noted.
 #
 class domain_membership (
-  $domain,
-  $username,
-  $password,
-  $secure_password = false,
-  $machine_ou      = undef,
-  $resetpw         = true,
-  $reboot          = true,
-  $reboot_apply    = 'finished',
-  $join_options    = '1',
-  $user_domain     = undef,
+  Stdlib::Fqdn $domain,
+  String $username,
+  Variant[String,Sensitive] $password,
+  Optional[Stdlib::Fqdn] $user_domain           = undef,
+  Boolean $secure_password                      = false,
+  String $machine_ou                            = '$null',
+  Boolean $resetpw                              = true,
+  Boolean $reboot                               = true,
+  Pattern[/immediately|finished/] $reboot_apply = 'finished',
```

Comment:
When choosing a list of possible Strings, the `Enum` data type is a bit cleaner and gives better error messages.

`Enum['immediately','finished']  $reboot_apply = 'finished',`

Reply:
Oh yeah! Thanks for the reminder.